In [6]:
!pip install langchain-google-genai
!pip install instructorembedding
!pip install huggingface
!pip install --upgrade huggingface_hub
!pip install -U sentence-transformers

DEPRECATION: pyodbc 4.0.0-unsupported has a non-standard version number. pip 24.0 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pyodbc or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063


   ---------------------------------------- 146.9/146.9 kB 2.2 MB/s eta 0:00:00
   ---------------------------------------- 598.7/598.7 kB 2.1 MB/s eta 0:00:00
   ---------------------------------------- 122.0/122.0 kB 2.4 MB/s eta 0:00:00
   ---------------------------------------- 184.2/184.2 kB 3.7 MB/s eta 0:00:00
   ---------------------------------------- 413.4/413.4 kB 3.7 MB/s eta 0:00:00
   ---------------------------------------- 228.7/228.7 kB 4.6 MB/s eta 0:00:00
   ---------------------------------------- 48.8/48.8 kB ? eta 0:00:00
   ---------------------------------------- 3.7/3.7 MB 3.7 MB/s eta 0:00:00
  Attempting uninstall: protobuf
    Found existing installation: protobuf 3.19.1
    Uninstalling protobuf-3.19.1:
      Successfully uninstalled protobuf-3.19.1
  Attempting uninstall: grpcio
    Found existing installation: grpcio 1.43.0
    Uninstalling grpcio-1.43.0:
      Successfully uninstalled grpcio-1.43.0
  Attempting uninstall: google-auth
    Found existing 

In [1]:
from langchain_google_genai import GoogleGenerativeAI
#Use langchain_google_genai.GoogleGenerativeAI instead.
api_key = "AIzaSyCC2pFam7_gLQY_MBNEMu8Zpjlwi02t578"
llm = GoogleGenerativeAI(model="models/text-bison-001", google_api_key=api_key)

In [2]:
poem = llm("Write a 4 line poem of my love for samosa")
print(poem)

**Samosas, oh samosas**
**So crispy, so flaky, so delicious**
**I could eat them all day long**
**My one true love**


In [3]:
from langchain.chains import RetrievalQA
from langchain.embeddings import GooglePalmEmbeddings
from langchain.llms import GooglePalm

### loading data from Codebasics FAQ csv file

In [5]:
from langchain.document_loaders.csv_loader import CSVLoader

loader = CSVLoader(file_path='codebasics_faqs.csv', source_column="prompt")

# Store the loaded data in the 'data' variable
data = loader.load()

### Hugging Face Embeddings

In [27]:
from langchain.embeddings import HuggingFaceInstructEmbeddings

In [32]:
from langchain.embeddings import HuggingFaceInstructEmbeddings
# Initialize instructor embeddings using the Hugging Face model
instructor_embeddings = HuggingFaceInstructEmbeddings(model_name="hkunlp/instructor-large")
e = instructor_embeddings.embed_query("What is your refund policy?")

load INSTRUCTOR_Transformer
max_seq_length  512


In [33]:
len(e)

768

In [34]:
e[:5]

[-0.043898068368434906,
 0.007685533724725246,
 -0.009231905452907085,
 0.024496253579854965,
 0.03359227627515793]

As you can see above, embedding for a sentance "What is your refund policy" is a list of size 768. Looking at the numbers in this list, doesn't give any intuitive understanding of what it is but just assume that these numbers are capturing the meaning of "What is your refund policy". If you are curious to know about embeddings, go to youtube and search "codebasics word embeddings" and you will find bunch of videos with simple, intuitive explanations

### Vector store using FAISS

In [ ]:
from langchain.vectorstores import FAISS

# Create a FAISS instance for vector database from 'data'
vectordb = FAISS.from_documents(documents=data,
                                 embedding=instructor_embeddings)

# Create a retriever for querying the vector database
retriever = vectordb.as_retriever(score_threshold = 0.7)

In [ ]:
rdocs = retriever.get_relevant_documents("how about job placement support?")
rdocs

### Create RetrievalQA chain along with prompt template 🚀

In [ ]:
from langchain.prompts import PromptTemplate

prompt_template = """Given the following context and a question, generate an answer based on this context only.
In the answer try to provide as much text as possible from "response" section in the source document context without making much changes.
If the answer is not found in the context, kindly state "I don't know." Don't try to make up an answer.

CONTEXT: {context}

QUESTION: {question}"""


PROMPT = PromptTemplate(
    template=prompt_template, input_variables=["context", "question"]
)
chain_type_kwargs = {"prompt": PROMPT}


from langchain.chains import RetrievalQA

chain = RetrievalQA.from_chain_type(llm=llm,
                            chain_type="stuff",
                            retriever=retriever,
                            input_key="query",
                            return_source_documents=True,
                            chain_type_kwargs=chain_type_kwargs)


In [ ]:
chain('Do you provide job assistance and also do you provide job gurantee?')